### PR TITLE
Add Cylinder Domain

### LINKED ISSUE #421: Benchmark Wedge2D
# Feature request:

Benchmark the functions in Wedge2D (as an exercise in learning how to benchmark code)

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Detailed discussion:

Test whether moving `std::move(physical_x *=-1)` will improve the code

### PR DESCRIPTION
**Pull request by mar-celine**
## Proposed changes

A useful Domain for testing.
Also revamps Wedge2D to be more similar in functionality to Wedge3D
This closes #421 and #422 

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### CODE REVIEW

**Comment by kidder** [comment]
@gsb76 fix, squash, and remove in progress label before we look...

File: src/Domain/CoordinateMaps/Wedge2D.hpp [hunk 1]
```diff
@@ -20,10 +20,14 @@ class er;
...
-/// flat face which is circumscribed by a circular arc of radius
-/// `radius_of_other`. These arcs extend \f$\pi/2\f$ in angle, and can be
+/// Two dimensional map from the logical square to a wedge, which is used by
+/// domains which use disks or annuli. The coordinate map is constructed by
+/// linearly interpolating between a bulged arc which is circumscribed by a
+/// circular arc of radius `radius_outer` and a bulged arc which is
+/// circumscribed by a circular arc of radius `radius_inner`. These arcs can be
+/// made to be straight or circular, based on the value of `circularity_inner`
+/// or `circularity_outer`, which can take on values between 0 (for straight)
+/// and 1 (for circular). These arcs extend \f$\pi/2\f$ in angle, and can be
```
Comment (kidder) [inline src/Domain/CoordinateMaps/Wedge2D.hpp]: explain how to get from the 3d case to the 2d case

File: src/Domain/CoordinateMaps/Wedge2D.cpp [hunk 2]
```diff
@@ -16,18 +16,35 @@
...
-             radius_of_other > sqrt(2.0) * radius_of_circle,
-         "The faces of the wedge must not intersect.");
+  ASSERT(radius_inner > 0.0, "This radius must be greater than zero.");
+  ASSERT(radius_outer > 0.0, "This radius must be greater than zero.");
+  ASSERT(radius_outer *
+                 ((1.0 - circularity_outer) / sqrt(2.0) + circularity_outer) >
+             radius_inner *
+                 ((1.0 - circularity_inner) / sqrt(2.0) + circularity_inner),
+         "The arguments passed into the constructor for Wedge2D result in an "
+         "object where the "
```
Comment (kidder) [inline src/Domain/CoordinateMaps/Wedge2D.cpp]: formatting

File: src/Domain/CoordinateMaps/Wedge2D.cpp [hunk 3]
```diff
@@ -103,10 +120,9 @@ std::array<tt::remove_cvref_wrap_t<T>, 2> Wedge2D::inverse(
 
   const ReturnType cap_eta = physical_y / physical_x;
   const ReturnType one_over_rho = 1.0 / sqrt(1.0 + square(cap_eta));
-  const double common_factor_one = 0.5 * radius_of_other_ / sqrt(2.0);
-  const ReturnType common_factor_two = 0.5 * radius_of_circle_ * one_over_rho;
-  ReturnType xi = (physical_x - (common_factor_two + common_factor_one)) /
-                  (common_factor_two - common_factor_one);
+  const ReturnType x_0 = scaled_trapezoid_zero_ + annulus_zero_ * one_over_rho;
+  const ReturnType x_1 = scaled_trapezoid_rate_ + annulus_rate_ * one_over_rho;
+  ReturnType xi = (physical_x - x_0) / x_1;
```
Comment (kidder) [inline src/Domain/CoordinateMaps/Wedge2D.cpp]: inline x_0 and x_1

File: src/Domain/CoordinateMaps/Wedge2D.cpp [hunk 4]
```diff
@@ -172,17 +187,29 @@ tnsr::Ij<tt::remove_cvref_wrap_t<T>, 2, Frame::NoFrame> Wedge2D::inv_jacobian(
...
+  p | annulus_zero_;
+  p | scaled_trapezoid_rate_;
+  p | annulus_rate_;
 }
 
 bool operator==(const Wedge2D& lhs, const Wedge2D& rhs) noexcept {
-  return lhs.radius_of_other_ == rhs.radius_of_other_ and
-         lhs.radius_of_circle_ == rhs.radius_of_circle_ and
+  return lhs.radius_inner_ == rhs.radius_inner_ and
+         lhs.radius_outer_ == rhs.radius_outer_ and
```
Comment (kidder) [inline src/Domain/CoordinateMaps/Wedge2D.cpp]: uh oh -Gabe

File: src/Domain/DomainCreators/Cylinder.hpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+/// \cond
+template <size_t Dim, typename Frame>
+class DomainCreator;  // IWYU pragma: keep
+/// \endcond
+
+namespace DomainCreators {
+
+/// \ingroup DomainCreatorsGroup
+/// Create a 3D Domain in the shape of a cylinder from a square prism
+/// surrounded by four wedge2d prisms.
```
Comment (kidder) [inline src/Domain/DomainCreators/Cylinder.hpp]: better description than wedge2d prisms (maybe a picture)

File: src/Domain/DomainCreators/Cylinder.cpp [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+          std::move(initial_refinement)),              // NOLINT
+      initial_number_of_grid_points_(                  // NOLINT
+          std::move(initial_number_of_grid_points)),   // NOLINT
+      use_equiangular_map_(use_equiangular_map) {}     // NOLINT
+
+template <typename TargetFrame>
+Domain<3, TargetFrame> Cylinder<TargetFrame>::create_domain() const noexcept {
+  using Affine = CoordinateMaps::Affine;
+  using Affine2D = CoordinateMaps::ProductOf2Maps<Affine, Affine>;
+  using Affine3DPrism = CoordinateMaps::ProductOf2Maps<Affine2D, Affine>;
```
Comment (kidder) [inline src/Domain/DomainCreators/Cylinder.cpp]: use ProductOf3Maps

File: src/Domain/DomainCreators/Cylinder.cpp [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+  using Equiangular = CoordinateMaps::Equiangular;
+  using Equiangular2D =
+      CoordinateMaps::ProductOf2Maps<Equiangular, Equiangular>;
+  using Equiangular3DPrism =
+      CoordinateMaps::ProductOf2Maps<Equiangular2D, Affine>;
```
Comment (kidder) [inline src/Domain/DomainCreators/Cylinder.cpp]: again

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/Domain/CoordinateMaps/Wedge2D.hpp [hunk 8]
```diff
@@ -20,19 +20,55 @@ class er;
...
+ * \end{bmatrix}\f]
+ *
+ * We will define the variables \f$T(\xi)\f$ and \f$A(\xi)\f$, the trapezoid
+ * and annulus factors: \f[T(\xi) = T_0 + T_1\xi\f] \f[A(\xi) = A_0 +
+ * A_1\xi\f]
+ * Where \f{align*}T_0 &= \frac{1}{2} \big\{ (1-c_{outer})R_{outer} +
+ * (1-c_{inner})R_{inner}\big\}\\
+ * T_1 &= \partial_{\xi} T = \frac{1}{2} \big\{ (1-c_{outer})R_{outer} -
+ * (1-c_{inner})R_{inner}\big\}\\
+ * A_0 &= \frac{1}{2} \big\{ s_{outer}R_{outer} + c_{inner}R_{inner}\big\}\\
```
Comment (kidder) [inline src/Domain/CoordinateMaps/Wedge2D.hpp]: s to c

File: src/Domain/CoordinateMaps/Wedge2D.hpp [hunk 9]
```diff
@@ -20,19 +20,55 @@ class er;
...
+ * and 1 (for circular). These arcs extend \f$\pi/2\f$ in angle, and can be
+ * oriented along the +/- x or y axis. The choice of using either equiangular
+ * or equidistant coordinates along the arcs is specifiable with
+ * `with_equiangular_map`. For a more detailed discussion, see the
+ * documentation for Wedge3D, where the default logical coordinate that points
+ * in the radial direction is zeta, and one may set either of the two other
+ * logical coordinates to zero to obtain an equivalent Wedge2D map. In terms
+ * of the variables defined in Wedge2D, the mapping is:
+ *
+ *  \f[\vec{x}(\xi,\eta) =
```
Comment (kidder) [inline src/Domain/CoordinateMaps/Wedge2D.hpp]: define rho and H

File: src/Domain/CoordinateMaps/Wedge2D.hpp [hunk 10]
```diff
@@ -19,21 +20,63 @@ class er;
...
+ * made to be straight or circular, based on the value of `circularity_inner`
+ * or `circularity_outer`, which can take on values between 0 (for straight)
+ * and 1 (for circular). These arcs extend \f$\pi/2\f$ in angle, and can be
+ * oriented along the +/- x or y axis. The choice of using either equiangular
+ * or equidistant coordinates along the arcs is specifiable with
+ * `with_equiangular_map`. For a more detailed discussion, see the
+ * documentation for Wedge3D, where the default logical coordinate that points
+ * in the radial direction is zeta, and one may set either of the two other
+ * logical coordinates to zero to obtain an equivalent Wedge2D map. In terms
+ * of the variables defined in Wedge2D (the circularity \f$c\f$ and the radius
```
Comment (kidder) [inline src/Domain/CoordinateMaps/Wedge2D.hpp]: In terms of the circularity ...

File: src/Domain/CoordinateMaps/Wedge2D.hpp [hunk 11]
```diff
@@ -19,21 +20,63 @@ class er;
...
+ * \f$R\f$), the mapping is:
+ *
+ *  \f[\vec{x}(\xi,\eta) =
```
Comment (kidder) [inline src/Domain/CoordinateMaps/Wedge2D.hpp]: define H and rho

File: src/Domain/DomainCreators/Cylinder.hpp [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+/// \cond
+template <size_t Dim, typename Frame>
+class DomainCreator;  // IWYU pragma: keep
+/// \endcond
+
+namespace DomainCreators {
+
+/// \ingroup DomainCreatorsGroup
+/// Create a 3D Domain in the shape of a cylinder where the cross-section
+/// is a square surrounded by four Wedge2Ds.
```
Comment (kidder) [inline src/Domain/DomainCreators/Cylinder.hpp]: surrounded by four two-dimensional wedges (see ref to Wedge2D)

File: src/Domain/CoordinateMaps/Wedge2D.hpp [hunk 13]
```diff
@@ -19,21 +20,63 @@ class er;
...
+ * circular arc of radius `radius_outer` and a bulged arc which is
+ * circumscribed by a circular arc of radius `radius_inner`. These arcs can be
+ * made to be straight or circular, based on the value of `circularity_inner`
+ * or `circularity_outer`, which can take on values between 0 (for straight)
+ * and 1 (for circular). These arcs extend \f$\pi/2\f$ in angle, and can be
+ * oriented along the +/- x or y axis. The choice of using either equiangular
+ * or equidistant coordinates along the arcs is specifiable with
+ * `with_equiangular_map`. For a more detailed discussion, see the
+ * documentation for Wedge3D, where the default logical coordinate that points
+ * in the radial direction is zeta, and one may set either of the two other
```
Comment (kidder) [inline src/Domain/CoordinateMaps/Wedge2D.hpp]: \zeta and mention that \xi is the radial direction here

**Review by kidder — CHANGES_REQUESTED** [review]
squash
Group review by Nils D, Will, Larry

File: src/Domain/CoordinateMaps/Wedge2D.cpp [hunk 14]
```diff
@@ -42,14 +42,14 @@ Wedge2D::Wedge2D(double radius_inner, double radius_outer,
...
                             (1.0 - circularity_inner) * radius_inner);
-  annulus_zero_ = 0.5 * (circularity_outer_ * radius_outer +
+  annulus_zero_ = 0.5 * (circularity_outer * radius_outer +
                          circularity_inner * radius_inner);
   scaled_trapezoid_rate_ = 0.5 / sqrt(2.0) *
-                           ((1.0 - circularity_outer_) * radius_outer -
+                           ((1.0 - circularity_outer) * radius_outer -
                             (1.0 - circularity_inner) * radius_inner);
-  annulus_rate_ = 0.5 * (circularity_outer_ * radius_outer -
+  annulus_rate_ = 0.5 * (circularity_outer * radius_outer -
```
Comment (kidder) [inline src/Domain/CoordinateMaps/Wedge2D.cpp]: move these fixes to a different commit

**Review by kidder — CHANGES_REQUESTED** [review]

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**